File: python/intro.py
student_data = [
    {'name': 'Bob', 'id': 0, 'scores':[68, 75, 56, 81]},
    {'name': 'Alice', 'id': 1, 'scores':[75, 90, 64, 88]},
    {'name': 'Carol', 'id': 2, 'scores':[59, 74, 71, 68]},
    {'name': 'Dan', 'id': 3, 'scores':[64, 58, 53, 62]},
]

def process_student_data(data, pass_threshold=60, merit_threshold=75):
    """ egy egyszerű statisztikát hoz létre a tanulók adataiból """
    # a fenti doc-string automatikusan hozzárendelődik a __doc__ attribútumhoz.

    for sdata in data:
        av = sum(sdata['scores']) / float(len(sdata['scores']))
        sdata['average'] = av

        if av > merit_threshold:
            sdata['assessment'] = 'passed with merit'
        elif av > pass_threshold:
            sdata['assessment'] = 'passed'
        else:
            sdata['assessment'] = 'failed'

        print("{}'s (id: {}) final asessment is {}.".format(sdata['name'], sdata['id'], sdata['assessment'].upper()))


if __name__ == '__main__':
    process_student_data(student_data)

# bármikor amikor a Python interpreterje értelmez egy forrásállományt,
# két dolgoz tesz:
#   - beállít néhány speciális változót, mint pl. a 'name',
#   - lefuttatja a fájlban található összes utasítást.
# a name speciális változó értéke akkor lesz 'main', ha elsődleges forrásként futtassuk.

#########################################

import logging

logger = logging.getLogger(__name__)

logger.debug('Some useful debugging output')
logger.info('Some general information')
logger.warning('Some warning')

#########################################

foo = {'first_key': 1, 'second_key': 2}

for item in foo:
    print(item)

# Ebben az esetben nem a dict itemet fogja kiírna, hanem annak kulcsát.

for item in foo.items():
    print(item)

# Ebben az esetben már a dict itemet fogja kiírni key-value páronként, tuple típusban. 

######################################### 

class Citizen(object):
    def __init__(self, name, country):
        self.name = name
        self.country = country

    def print_details(self):
        print('Citizen {} from {}.'.format(self.name, self.country))

c = Citizen('Lukács Zs.', 'Hungary')
c.print_details()

######################################### 

class Winner(Citizen):
    def __init__(self, name, country, category, year):
        super(Winner, self).__init__(name, country)
        self.category = category
        self.year = year

    def print_details(self):
        print('Nobel winner {} from {}, category {}, year {}'.format(self.name, self.country, self.category, self.year))

w = Winner('Albert E.', 'Switzerland', 'Physics', 1921)
w.print_details()

#########################################

names = ['Alice', 'Bob', 'Carol']

for i, n in enumerate(names):
    print('{}: {}'.format(i, n))

#########################################

def fibonacci(n):
    x, y = 0, 1

    for i in range(n):
        print(x)
        x, y = y, x + y

fibonacci(4)

#########################################

from collections import Counter

items = ['F', 'B', 'B', 'F', 'A', 'C', 'D', 'D']

cntr = Counter(items)
# a Counter függvény megszámolja az elemek gyakoriságát

print(cntr)

File: javascript/intro.js
"use strict"

let studentData = [{
        'name': 'Bob',
        'id': 0,
        'scores': [68, 75, 56, 81]
    },
    {
        'name': 'Alice',
        'id': 1,
        'scores': [75, 90, 64, 88]
    },
    {
        'name': 'Carol',
        'id': 2,
        'scores': [59, 74, 71, 68]
    },
    {
        'name': 'Dan',
        'id': 3,
        'scores': [64, 58, 53, 62]
    },
]

function processStudentData(data, passThreshold, meritThreshold) {
    passThreshold = typeof passThreshold !== 'undefined' ? passThreshold : 60;
    meritThreshold = typeof meritThreshold !== 'undefined' ? meritThreshold : 75;
    /* A fenti kód esetében inicializálja a paramétereket, 
    amennyiben azok nem kaptak kezdeti értéket a függvény hívásakor */

    data.forEach((sdata) => {
        let av = sdata.scores.reduce((prev, current) => prev + current, 0) / sdata.scores.length;
        sdata.average = av;

        if (av > meritThreshold) {
            sdata.assessment = 'passed with merit';
        } else if (av > passThreshold) {
            sdata.assessment = 'passed';
        } else {
            sdata.assessment = 'failed';
        }

        console.log(sdata.name + "'s (id: " + sdata.id + ") final assessment is: " + sdata.assessment.toUpperCase());
    });
}

processStudentData(studentData);

///////////////////////////////

let foo = {
    'firstKey': 1,
    'secondKey': 2
};

for (let item in foo) {
    if (foo.hasOwnProperty(item)) {
        console.log(item + " " + foo[item]);
    }
}

///////////////////////////////

let Citizen = function(name, country) {
    this.name = name;
    this.country = country;
}

Citizen.prototype = {
    printDetails: function() {
        console.log('Citizen ' + this.name + ' from ' + this.country);
    }
}

let c = new Citizen('Lukács Zs.', 'Hungary');

c.printDetails();

///////////////////////////////

let Citizen = {
    setCitizen: function(name, country) {
        this.name = name;
        this.country = country;
        return this;
    },
    printDetails: function() {
        console.log('Citizen ' + this.name + ' from ' + this.country);
    }
}

let Winner = Object.create(Citizen);

Winner.setWinner = function(name, country, category, year) {
    this.setCitizen(name, country);
    this.category = category;
    this.year = year;
    return this;
}

Winner.printDetails = function() {
    console.log('Nobel winner ' + this.name + ' from ' + this.country + ', category ' + this.category + ', year ' + this.year);
}

let a = Object.create(Winner).setWinner('Einstein A.', 'Switzerland', 'Physics', '1921');

a.printDetails();

///////////////////////////////

let names = ['Alice', 'Bob', 'Carol']

names.forEach(function(i, n) {
    console.log(i + ': ' + n);
})

///////////////////////////////

let items = ['F', 'B', 'B', 'F', 'A', 'C', 'D', 'D'];

_.countBy(items);

///////////////////////////////

function Counter(inc) {
    let count = 0;
    let add = function() {
        count += inc;
        console.log('Current count: ' + count);
    }
    return add;
}

let cntr = Counter(2);

cntr();

cntr();

///////////////////////////////

function outer(bar) {
    this.bar = bar;
    var that = this;
    function inner(baz) {
        this.baz = baz * that.bar;
        // ...
        // ebben az esetben a that a külső függvény this-jére vonatkozik.
    }
}